File: android/app/src/main/java/com/pigeon/mobile/auth/AuthError.kt
package com.pigeon.mobile.auth

import uniffi.pigeon_mobile_core.CoreException
import uniffi.pigeon_mobile_core.ErrorCode

/**
 * Maps a typed [CoreException] to a user-facing message.
 *
 * The UI branches on the typed code, never on error text (CLAUDE.md) — and it
 * must handle every variant, because a federated, offline-prone client will hit
 * them. A pure function so it can be unit-tested without a device.
 */
fun authErrorMessage(e: CoreException): String = when (e) {
    is CoreException.Api -> when (e.code) {
        is ErrorCode.UserInUse -> "That username is already taken."
        is ErrorCode.Forbidden -> "Incorrect username or password."
        is ErrorCode.InvalidUsername -> "Usernames may use only a–z, 0–9, and . _ -"
        is ErrorCode.UnknownToken, is ErrorCode.MissingToken ->
            "Your session has expired. Please sign in again."
        is ErrorCode.LimitExceeded -> "Too many attempts. Please wait a moment and try again."
        is ErrorCode.BadJson, is ErrorCode.NotJson -> "The server rejected the request."
        is ErrorCode.NotFound -> "Not found on this server."
        else -> "The server reported an error."
    }
    is CoreException.Network -> "Can't reach the server. Check the address and your connection."
    is CoreException.Protocol -> "Unexpected response from the server."
    is CoreException.Storage -> "Couldn't access secure storage on this device."
    is CoreException.Crypto -> "A security error occurred."
}

File: ios/Pigeon/Auth/AuthError.swift
// AuthError (M5.4) — maps a typed `CoreError` to a user-facing message. The iOS
// mirror of Android's `AuthError.kt`; it must stay in lockstep with it (same
// codes → same wording) so the two apps behave identically.
//
// The UI branches on the typed `P_` code, NEVER on error text (CLAUDE.md), and
// handles every variant — a federated, offline-prone client will hit them all.
// A pure function, unit-tested without a simulator (AuthErrorTests).
import PigeonCore

func authErrorMessage(_ error: Error) -> String {
    guard let core = error as? CoreError else {
        // Non-core errors shouldn't reach the UI from the core's async FFI, but
        // never show a raw Swift error — degrade to a generic message.
        return "Something went wrong."
    }
    switch core {
    case let .Api(code, _):
        switch code {
        case .userInUse: return "That username is already taken."
        case .forbidden: return "Incorrect username or password."
        case .invalidUsername: return "Usernames may use only a–z, 0–9, and . _ -"
        case .unknownToken, .missingToken:
            return "Your session has expired. Please sign in again."
        case .limitExceeded: return "Too many attempts. Please wait a moment and try again."
        case .badJson, .notJson: return "The server rejected the request."
        case .notFound: return "Not found on this server."
        case .badSignature, .unrecognized, .unknown, .other:
            return "The server reported an error."
        }
    case .Network: return "Can't reach the server. Check the address and your connection."
    case .Protocol: return "Unexpected response from the server."
    case .Storage: return "Couldn't access secure storage on this device."
    case .Crypto: return "A security error occurred."
    }
}
